packages/shared-types/src/CommonTypes.ts

```diff
@@ -53,6 +56,17 @@ export type Renderable<P = never> =
   | (() => ReactNode | Element)
   | Element
 
+/**
+ * Union type helper for the "children" prop.
+ * The argument has to be type of, or union of type of
+ * React.ComponentElement<any, any> */
+export type ChildrenOfType<T extends ComponentElement<any, any>> =
+  | T
+  | undefined
+  | null
+  | (T | undefined | null)[]
+  | (T | undefined | null | (T | undefined | null)[])[]
```

Comment:
If we don't care about restricting the depth of nesting of arrays, then we can rewrite this type as:

```ts
type ChildrenOfType<T extends ComponentElement<any, any>> =
| T 
| undefined
| null
| ChildrenOfType<T>[]
```

[Example](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gWzHAdhFwCiAbCD3AA8A0lBAB7DoAmAzlLcAE4CWKA5gDRQBqUAvFACGKEAD4BUAEoQAxnCbUSnHmICwAKAD0AKh2aoOqAFUULVFFCQoACwjZITKADMFlu1ABEsmy2zUmdE8oMCY4MAA6AyMAFQ8hJnYAV3wMWyF6YDgoACNoK2g4Z243JLMLIstwQudo6QghWWAI+CRUAhw8AkIREG5eiR0tTQpkJmAq6xhff0CUAHlnGOrCGLJKGnpW5DQMTtSiXv7RMQl+AygAHygYi+uy6ghnNghqO6gUJOxsd4AKNfuKEezzQ1CuHy+2AAlABtAC6fwBUAeTxeYOun2+4P+4JRINe4Mx0PhsIRGk0BVgMwC6EWy0gACZVusqECtogdh1cAceqJjuIzpprmshcigajQaKiaLpn4aQsltUmTExPDNOqNPIUIwoE0AFxUuVzOkrADenjs3zgngNnmAdkCngAvmcoDDTZbsHBbfaII6ndx3Z7vV5ff7uB77F6fQ6IM7uES4YGg1GQ3bY864VmNVqdU0GQbZbNaYrGYRzcGbaGMy7JCmrTG-XGA27Iw3q0341A29GO+GId8k26idw8WjOPXe+nO06s2SgA)

Reply:
It seems that React doesn't care about the depth, so I'll update it! Thanks, good tip!